File: clients/rn/__tests__/engine/room-state.test.ts
import { emptyRoomState, applyEvent, toSnapshot, RoomState } from '../../src/engine/room/RoomState';
import type { WorldEvent } from '../../src/engine/events/WorldEvent';

describe('RoomState', () => {
  const ts = Date.now();

  it('empty state has no entities, exits, or objects', () => {
    const state = emptyRoomState('test');
    expect(state.roomId).toBe('test');
    expect(state.name).toBe('');
    expect(Object.keys(state.entities)).toHaveLength(0);
    expect(Object.keys(state.exits)).toHaveLength(0);
    expect(Object.keys(state.objects)).toHaveLength(0);
  });

  it('applies room_created', () => {
    const state = applyEvent(emptyRoomState('test'), {
      type: 'room_created', roomId: 'test', timestamp: ts,
      name: 'The Nexus', description: 'A hub.', zone: 'foundation',
    });
    expect(state.name).toBe('The Nexus');
    expect(state.description).toBe('A hub.');
    expect(state.zone).toBe('foundation');
  });

  it('applies entity_entered and entity_left', () => {
    let state = emptyRoomState('test');
    state = applyEvent(state, {
      type: 'entity_entered', roomId: 'test', timestamp: ts,
      entityId: 'p1', entityName: 'Alice', entityType: 'player', fromDirection: 'south',
    });
    expect(state.entities['p1']).toBeDefined();
    expect(state.entities['p1'].name).toBe('Alice');

    state = applyEvent(state, {
      type: 'entity_left', roomId: 'test', timestamp: ts,
      entityId: 'p1', entityName: 'Alice', direction: 'north',
    });
    expect(state.entities['p1']).toBeUndefined();
  });

  it('applies object_added and object_taken', () => {
    let state = emptyRoomState('test');
    state = applyEvent(state, {
      type: 'object_added', roomId: 'test', timestamp: ts,
      objectId: 'o1', objectName: 'sword', description: 'A blade.', takeable: true,
    });
    expect(state.objects['o1']).toBeDefined();
    expect(state.objects['o1'].name).toBe('sword');
    expect(state.objects['o1'].takeable).toBe(true);

    state = applyEvent(state, {
      type: 'object_taken', roomId: 'test', timestamp: ts,
      entityId: 'p1', objectId: 'o1', objectName: 'sword',
    });
    expect(state.objects['o1']).toBeUndefined();
  });

  it('applies exit_opened and exit_closed', () => {
    let state = emptyRoomState('test');
    state = applyEvent(state, {
      type: 'exit_opened', roomId: 'test', timestamp: ts,
      direction: 'north', targetRoom: 'terminal', label: 'To Terminal',
    });
    expect(state.exits['north']).toBeDefined();
    expect(state.exits['north'].targetRoom).toBe('terminal');

    state = applyEvent(state, {
      type: 'exit_closed', roomId: 'test', timestamp: ts, direction: 'north',
    });
    expect(state.exits['north']).toBeUndefined();
  });

  it('applies description_changed', () => {
    let state = applyEvent(emptyRoomState('test'), {
      type: 'room_created', roomId: 'test', timestamp: ts,
      name: 'Room', description: 'Old desc.', zone: 'z',
    });
    state = applyEvent(state, {
      type: 'description_changed', roomId: 'test', timestamp: ts,
      newDescription: 'New desc.', reason: 'script',
    });
    expect(state.description).toBe('New desc.');
  });

  it('applies hints_updated', () => {
    let state = emptyRoomState('test');
    state = applyEvent(state, {
      type: 'hints_updated', roomId: 'test', timestamp: ts,
      hints: [{ label: 'Say hi', intent: 'greet', action: 'say' }],
    });
    expect(state.hints).toHaveLength(1);
    expect(state.hints[0].label).toBe('Say hi');
  });

  it('applies property_changed (set and remove)', () => {
    let state = emptyRoomState('test');
    state = applyEvent(state, {
      type: 'property_changed', roomId: 'test', timestamp: ts,
      key: 'mood', oldValue: null, newValue: 'happy',
    });
    expect(state.properties['mood']).toBe('happy');

    state = applyEvent(state, {
      type: 'property_changed', roomId: 'test', timestamp: ts,
      key: 'mood', oldValue: 'happy', newValue: null,
    });
    expect(state.properties['mood']).toBeUndefined();
  });

  it('said event does not change state', () => {
    const state = emptyRoomState('test');
    const after = applyEvent(state, {
      type: 'said', roomId: 'test', timestamp: ts,
      entityId: 'p1', entityName: 'Alice', text: 'Hello',
    });
    expect(after).toBe(state);
  });

  it('toSnapshot produces correct wire format', () => {
    let state = emptyRoomState('test');
    state = applyEvent(state, {
      type: 'room_created', roomId: 'test', timestamp: ts,
      name: 'Room', description: 'Desc.', zone: 'z',
    });
    state = applyEvent(state, {
      type: 'entity_entered', roomId: 'test', timestamp: ts,
      entityId: 'p1', entityName: 'Alice', entityType: 'player', fromDirection: 'south',
    });
    const snapshot = toSnapshot(state);
    expect(snapshot.roomId).toBe('test');
    expect(snapshot.name).toBe('Room');
    expect(snapshot.entities).toHaveLength(1);
    expect(snapshot.entities[0].name).toBe('Alice');
  });
});

File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/room/RoomStateTest.kt
package org.wyrdsekai.app.engine.room

import kotlin.time.Clock
import org.wyrdsekai.app.engine.event.WorldEvent
import org.wyrdsekai.app.protocol.*
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class RoomStateTest {
    private val now = Clock.System.now()
    private val roomId = "test-room"

    @Test
    fun emptyState() {
        val state = RoomState.empty(roomId)
        assertEquals(roomId, state.roomId)
        assertEquals("", state.name)
        assertTrue(state.exits.isEmpty())
        assertTrue(state.entities.isEmpty())
        assertTrue(state.objects.isEmpty())
    }

    @Test
    fun applyRoomCreated() {
        val state = RoomState.empty(roomId)
            .apply(WorldEvent.RoomCreated(roomId, now, "The Nexus", "A hub.", "foundation"))
        assertEquals("The Nexus", state.name)
        assertEquals("A hub.", state.description)
        assertEquals("foundation", state.zone)
    }

    @Test
    fun applyEntityEnteredAndLeft() {
        var state = RoomState.empty(roomId)
            .apply(WorldEvent.EntityEntered(roomId, now, "player-1", "Alice", "player", "north"))
        assertEquals(1, state.entities.size)
        assertEquals("Alice", state.entities["player-1"]!!.name)

        state = state.apply(WorldEvent.EntityLeft(roomId, now, "player-1", "Alice", "south"))
        assertTrue(state.entities.isEmpty())
    }

    @Test
    fun applyObjectTakenAndDropped() {
        var state = RoomState.empty(roomId)
            .apply(WorldEvent.ObjectAdded(roomId, now, "obj-1", "sword", "A sharp sword", true))
        assertEquals(1, state.objects.size)
        assertEquals("sword", state.objects["obj-1"]!!.name)

        state = state.apply(WorldEvent.ObjectTaken(roomId, now, "player-1", "obj-1", "sword"))
        assertTrue(state.objects.isEmpty())

        state = state.apply(WorldEvent.ObjectDropped(roomId, now, "player-1", "obj-1", "sword", "A sharp sword", true))
        assertEquals(1, state.objects.size)
    }

    @Test
    fun applyExitOpenedAndClosed() {
        var state = RoomState.empty(roomId)
            .apply(WorldEvent.ExitOpened(roomId, now, "north", "room-2", "A door"))
        assertEquals(1, state.exits.size)
        assertEquals("room-2", state.exits["north"]!!.targetRoom)

        state = state.apply(WorldEvent.ExitClosed(roomId, now, "north"))
        assertTrue(state.exits.isEmpty())
    }

    @Test
    fun applyDescriptionChanged() {
        val state = RoomState.empty(roomId)
            .apply(WorldEvent.RoomCreated(roomId, now, "Room", "Old desc", "zone"))
            .apply(WorldEvent.DescriptionChanged(roomId, now, "New desc", "script"))
        assertEquals("New desc", state.description)
    }

    @Test
    fun applyHintsUpdated() {
        val hints = listOf(Hint("Try this", "hint1", "say", null))
        val state = RoomState.empty(roomId)
            .apply(WorldEvent.HintsUpdated(roomId, now, hints))
        assertEquals(1, state.hints.size)
        assertEquals("Try this", state.hints[0].label)
    }

    @Test
    fun applyPropertyChanged() {
        var state = RoomState.empty(roomId)
            .apply(WorldEvent.PropertyChanged(roomId, now, "key1", null, "value1"))
        assertEquals("value1", state.properties["key1"])

        state = state.apply(WorldEvent.PropertyChanged(roomId, now, "key1", "value1", null))
        assertNull(state.properties["key1"])
    }

    @Test
    fun saidDoesNotChangeState() {
        val state = RoomState.empty(roomId)
            .apply(WorldEvent.RoomCreated(roomId, now, "Room", "Desc", "zone"))
        val after = state.apply(WorldEvent.Said(roomId, now, "p1", "Alice", "Hello"))
        assertEquals(state, after)
    }

    @Test
    fun toSnapshot() {
        val state = RoomState.empty(roomId)
            .apply(WorldEvent.RoomCreated(roomId, now, "The Nexus", "A hub.", "foundation"))
            .apply(WorldEvent.ExitOpened(roomId, now, "north", "terminal", "North"))
            .apply(WorldEvent.EntityEntered(roomId, now, "p1", "Alice", "player", "south"))
        val snapshot = state.toSnapshot()
        assertEquals("The Nexus", snapshot.name)
        assertEquals(1, snapshot.exits.size)
        assertEquals(1, snapshot.entities.size)
    }

    @Test
    fun applyObjectUsed() {
        val state = RoomState.empty(roomId)
            .apply(WorldEvent.ObjectAdded(roomId, now, "obj-1", "crystal", "A crystal", false))
        val after = state.apply(WorldEvent.ObjectUsed(roomId, now, "p1", "obj-1", "crystal", null, null))
        assertEquals(state, after) // ObjectUsed doesn't change state
    }

    @Test
    fun applyWhispered() {
        val state = RoomState.empty(roomId)
        val after = state.apply(WorldEvent.Whispered(roomId, now, "p1", "Alice", "p2", "psst"))
        assertEquals(state, after) // Whispered doesn't change state
    }

    @Test
    fun applyScriptTriggered() {
        val state = RoomState.empty(roomId)
        val after = state.apply(WorldEvent.ScriptTriggered(roomId, now, "nexus", "onEnter", emptyMap()))
        assertEquals(state, after) // ScriptTriggered doesn't change state
    }
}
